doc/cmis-module-enhancement/cmis-module-enhancement.md

```diff
@@ -0,0 +1,224 @@
+* TX signal is enabled (via setting SAI_PORT_ATTR_HOST_TX_SIGNAL_ENABLE to TRUE)
+* Admin Status is set to UP
+
+Whenever one of them is set to false (e.g. TX signal is disabled due to removal event of Admin station is set to DOWN), the SDK/FW stops transmission of this signal.
+
+This approach allows to avoid the issue of unterminated transmission when ASIC starts the transmission of the high-speed signal even when no module is plugged-in causing cross-talks to adjacent ports, high EMI/BER and shortening the transceiver lifespan
+
+On port creation for platforms supporting proposed enhancements, the SAI_PORT_ATTR_HOST_TX_SIGNAL_ENABLE shall be explicitly set by Ports OA to FALSE since its default SAI value is defined as TRUE for backward-compatibility reasons.
+
+#### 6.3.3. Handling of HW-based “host_tx_ready” event
+Per description above, on the platform supporting new enhancements, the SDK/FW shall send an asynchronous notification of “host_tx_ready” to SWSS when the ASIC starts/stops sending high-speed signal to a module.
+
+The vendor SDK shall internally support this indication from ASIC (via the trap mechanism) and shall call the notification callback registered by Ports OA on initialization .
+The backward-compatibility is supported by a new design. Therefore, the handling shall be as follows:
+
+| Platform Type                            | Ports OA Behavior                                                                                                                                            |   |
+|------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------|---|
+| Platform supporting new enhancements     | On return from SAI call of setting port's Admin status, nothing is done, unless the notification event "host_tx_ready" is received and consumed by Ports OA. |   |
+| Platform not supporting new enhancements | Ports OA will set "host_tx_ready" in STATE DB right after Admin status SAI call returns.<br>(UP -> "true", DOWN -> "false")                                  |   |
+
+
+##### 6.3.3.1. Warm-boot handling
+After warm-boot, SONiC will not receive a new “host_tx_ready” notification from SAI (because nothing has changed in ASIC/module). 
+To have the update status in STATE DB, one needs to refresh the status of “host_tx_ready” like it is done today for port’s operational status and speed.
+To do that a new logic related to host_tx_ready will be added in PortsOrch::initializePort().
+For platforms supporting proposed enhancements, SAI_PORT_ATTR_HOST_TX_READY_STATUS SAI attribute shall be queried and then the “host_tx_ready” flag in the STATE DB shall be updated accordingly.
+
+```
+bool PortsOrch::initializePort(Port &port)
+{
+    ...
+    ...
+    if (saiTxReadyNotifySupported && saiHwTxSignalEnabled)
```

Comment:
@noaOrMlnx what about platform that does not support this hw notification mechanism? Can we have parity in behavior for both?

Reply:
@prgeor 
If a platform does not support notification mechanism, but does support the enable signal, then following situation can happen:
1. host tx ready will not set to true when admin status is set to up. 
2. "allow" signal will be sent to lower layers saying the module is plugged in. 
3. host_tx_ready will never be set to TRUE since a notification that high-speed signal has been transmitted will never arrive. 

That's the reason we need to make sure those 2 queries are supported.

Reply:
@noaOrMlnx I was referring to host_tx_ready which gets deleted from STATE_DB table after a warm-reboot. With your proposal, for platform that support notification mechanis, the host_tx_ready will be restored after warm-reboot. But for other platforms, host_tx_ready will remain deleted from STATE_DB table causing Xcvrd to bring the link down. See my previous comment above: We should not delete PORT_TABLE from STATE_DB table during warm-reboot that way we can keep the behavior same for all platforms.

Reply:
Hi @prgeor 
Can you explain why do you expect link-flap when host_tx_ready is deleted with the deletion of PORT_TABLE? It is the exact same behavior as before. 

IMO reserving PORT_TABLE will not be the right solution.  
In PORT_TABLE there are some other keys as 'oper_status'. We do not want to have this key reserved between boots, but we need to know the real status - if the port link is up or not.
That's why we get the real oper status after boot is done via query to SAI.

Reply:
@noaOrMlnx oper_status i agree. I was talking about `host_tx_ready`. Have you tested if PORT_TABLE's `host_tx_ready` is not deleted ? The impact of host_tx_ready deleted during warm-reboot is Xcvrd re-initializing the CMIS optics causing link flap

Reply:
It is not deleted, but host_tx_ready value turns to false